core/include/scipp/core/element_array.h

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2019 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#ifndef SCIPP_CORE_ELEMENT_ARRAY_H
+#define SCIPP_CORE_ELEMENT_ARRAY_H
+
+#include <algorithm>
+#include <memory>
+
+#include "scipp/common/index.h"
+#include "scipp/core/aligned_allocator.h"
+
+namespace scipp::core::detail {
+
+/// Replacement for C++20 std::make_unique_default_init
+template <class T> auto make_unique_default_init(const scipp::index size) {
+  return std::unique_ptr<T>(new std::remove_extent_t<T>[size]);
+}
+
+static constexpr auto default_init_elements = []() {};
+using default_init_elements_t = decltype(default_init_elements);
+
+template <class T> class element_array {
+public:
+  using value_type = T;
+
+  element_array() noexcept = default;
+
+  explicit element_array(const scipp::index new_size, const T &value = T()) {
+    resize(new_size, default_init_elements);
+    std::fill(data(), data() + size(), value);
+  }
+
+  element_array(const scipp::index new_size, const default_init_elements_t &) {
+    resize(new_size, default_init_elements);
+  }
+
+  template <class InputIt,
+            std::enable_if_t<!std::is_integral<InputIt>{}, int> = 0>
+  element_array(InputIt first, InputIt last) {
+    resize(std::distance(first, last), default_init_elements);
+    std::copy(first, last, data());
+  }
+
+  element_array(std::initializer_list<T> init)
+      : element_array(init.begin(), init.end()) {}
+
+  element_array(element_array &&other) noexcept
+      : m_data(std::move(other.m_data)) {
+    m_size = other.m_size;
+    other.m_size = -1;
```

Comment:
Yes it is necessary to ensure that `other` is in a sensible state after the move. Without changing the `other.m_size` it will report a wrong size and users may access invalid memory.